src/utils.jl

```diff
@@ -133,3 +133,69 @@ function get_problem_data(model::MOI.AbstractOptimizer)
     
     return Q, q, G, h, A, b, nz, var_idx, nineq, ineq_con_idx, neq, eq_con_idx
 end
+
+
+"""
+    projection of vector `z` on MOI set `cone`
+"""
+function π(cone::MOI.AbstractVectorSet, z::Array{Float64})
+    if cone isa MOI.Zeros
+        return zeros(Float64, size(z))
+    elseif cone isa MOI.Nonnegatives
+        return max.(z,0.0)
+    elseif cone isa MOI.SecondOrderCone
+        t = z[1]
+        x = z[2:length(z)]
+        norm_x = norm(x)
+        if norm_x <= t
+            return copy(z)
+        elseif norm_x <= -t
+            return zeros(Float64, size(z))
+        else
+            result = zeros(Float64, size(z))
+            result[1] = 1.0
+            result[2:length(z)] = x / norm_x
+            result *= (norm_x + t) / 2.0
+            return result
+        end
+    end
+end
+
+"""
+    derivative of the projection of vector `z` on MOI set `cone`
+"""
+function Dπ(cone::MOI.AbstractVectorSet, z::Array{Float64})
```

Comment:
The function is type-unstable, with sometimes a matrix, sometimes a vector being returned

Reply:
Is it an issue if it's only the union of two types ?

Reply:
actually for `BlockDiagonals` to work, i need to turn vectors returned by this method into matrices

Reply:
will change it here only. that way the function will be type stable too